embassy-stm32/src/dma/ringbuffer.rs

```diff
@@ -54,151 +53,130 @@ pub trait DmaCtrl {
+                this is in a critical section to try to reduce mushy behavior.
+                it's not ideal but it's the best we can do
 
-            Ok(0)
-        } else if self.first < end {
-            // The available, unread portion in the ring buffer DOES NOT wrap
+                then, get the current position of of the dma write and check
+                if it's inside data we could have copied
+            */
+            let (pos, complete_count) =
+                critical_section::with(|_| (self.pos(dma.get_remaining_transfers()), dma.get_complete_count()));
+            if (complete_count > 0 && pos >= end) || complete_count > 1 || (pos >= self.start && pos < end) {
+                Err(OverrunError)
+            } else {
+                self.start = (self.start + len) % self.len();
 
-            if dma.get_complete_count() > 1 {
-                return Err(OverrunError);
+                Ok((len, self.len() - self.start))
             }
+        } else if self.start + buf.len() < self.len() {
+            // The available, unread portion in the ring buffer DOES wrap
+            // The DMA writer has wrapped since we last read and is currently
+            // writing (or the next byte added will be) in the beginning of the ring buffer.
 
-            // Copy out the bytes from the dma buffer
-            let len = self.copy_to(buf, self.first..end);
+            // The provided read buffer is not large enough to include all bytes from the tail of the dma buffer.
+
+            // Copy out from the dma buffer
+            let len = self.copy_to(buf, self.start..self.len());
 
             compiler_fence(Ordering::SeqCst);
 
-            match dma.get_complete_count() {
-                0 => {
-                    // The DMA writer has not wrapped before nor after the copy
-                }
-                1 => {
-                    // The DMA writer has written such that the ring buffer now wraps
-                    self.ndtr = dma.ndtr();
-                    if self.end() > self.first || dma.get_complete_count() > 1 {
-                        // The bytes that we have copied out have overflowed
-                        // as the writer has now both wrapped and is currently writing
-                        // within the region that we have just copied out
-                        return Err(OverrunError);
-                    }
-                }
-                _ => {
-                    return Err(OverrunError);
-                }
-            }
+            /*
+                first, check if the dma has wrapped around more than once
 
-            self.first = (self.first + len) % self.dma_buf.len();
-            Ok(len)
+                then, get the current position of of the dma write and check
+                if it's inside data we could have copied
+            */
+            let pos = self.pos(dma.get_remaining_transfers());
+            if dma.get_complete_count() > 1 || pos > self.start || pos < end {
```

Comment:
Again, what was wrong with `pos > self.start || dma.get_complete_count() > 1` like it was?
I think that `dma.get_complete_count() > 1` should always be the last condition tested in these or's, as that can change behind our back by the dma controller. The last condition `pos < end`, what does it capture?. That can only happen if `dma.get_complete_count() > 1` which should be tested as the last condition.

Reply:
You're right the `complete_count` should be the last condition tested.

> The last condition pos < end, what does it capture?. That can only happen if dma.get_complete_count() > 1 which should be tested as the last condition.

No, it can't. You can *never* assume that an interrupt will fire, ever. The most you can say is that, after some fixed amount of time, it is statistically certain to have fired, but it is never guaranteed. And these are practically free checks because these variables are already on the stack.

Reply:
The same goes here. I think it is ok to rely on `get_complete_count()` to actually return the correct value, as it can easily ensure that without relying on that the interrupt has actually fired.